src/Features/LanguageServer/Protocol/Handler/CodeActions/CodeActionsHandler.cs

```diff
@@ -35,81 +39,107 @@ internal class CodeActionsHandler : AbstractRequestHandler<LSP.CodeActionParams,
+                request.Range, cancellationToken).ConfigureAwait(false);
 
-        internal static async Task<IEnumerable<CodeAction>> GetCodeActionsAsync(Document? document,
-            ICodeFixService codeFixService,
-            ICodeRefactoringService codeRefactoringService,
-            LSP.Range selection,
-            CancellationToken cancellationToken)
-        {
-            if (document == null)
+            var codeFixes = codeFixCollections.SelectMany(c => c.Fixes);
+            using var _ = ArrayBuilder<VSCodeAction>.GetInstance(out var results);
+
+            // Go through code fixes and code refactorings separately so that we can properly set the CodeActionKind.
+            foreach (var codeFix in codeFixes)
             {
-                return ImmutableArray<CodeAction>.Empty;
+                // Filter out code actions with options since they'll show dialogs and we can't remote the UI and the options.
+                if (codeFix.Action is CodeActionWithOptions)
+                {
+                    continue;
+                }
+
+                // Temporarily filter out suppress and configure code actions, as we'll later combine them under a top-level
+                // code action.
+                if (codeFix.Action is AbstractConfigurationActionWithNestedActions)
+                {
+                    continue;
+                }
+
+                results.Add(GenerateVSCodeAction(request, codeFix.Action, CodeActionKind.QuickFix));
             }
 
-            var text = await document.GetTextAsync(cancellationToken).ConfigureAwait(false);
+            foreach (var (action, _) in codeRefactorings.SelectMany(codeRefactoring => codeRefactoring.CodeActions))
+            {
+                // Filter out code actions with options since they'll show dialogs and we can't remote the UI and the options.
+                if (action is CodeActionWithOptions)
+                {
+                    continue;
+                }
 
-            var textSpan = ProtocolConversions.RangeToTextSpan(selection, text);
-            var codeFixCollections = await codeFixService.GetFixesAsync(document, textSpan, true, cancellationToken).ConfigureAwait(false);
-            var codeRefactorings = await codeRefactoringService.GetRefactoringsAsync(document, textSpan, cancellationToken).ConfigureAwait(false);
+                results.Add(GenerateVSCodeAction(request, action, CodeActionKind.Refactor));
+            }
 
-            var codeActions = codeFixCollections.SelectMany(c => c.Fixes.Select(f => f.Action)).Concat(
-                                codeRefactorings.SelectMany(r => r.CodeActions.Select(ca => ca.action)));
+            // Special case (also dealt with specially in local Roslyn): 
+            // If we have configure/suppress code actions, combine them under one top-level code action.
+            var configureSuppressActions = codeFixes.Where(a => a.Action is AbstractConfigurationActionWithNestedActions);
```

Comment:
```suggestion
            var configureSuppressActions = codeFixes.OfType<AbstractConfigurationActionWithNestedActions>();
```

Reply:
Will this work since we wanted to check the type of `a.Action` and not just `a`?